**NumPy-> Numerical Python**

`NumPy is a Library which is used to create a data in the form of Dimension or work with the data which is already in Dimension`


In [1]:
import numpy as np

In [ ]:
dir(np)

['False_',
 'ScalarType',
 'True_',
 '_CopyMode',
 '_NoValue',
 '__NUMPY_SETUP__',
 '__all__',
 '__array_api_version__',
 '__builtins__',
 '__cached__',
 '__config__',
 '__dir__',
 '__doc__',
 '__expired_attributes__',
 '__file__',
 '__former_attrs__',
 '__future_scalars__',
 '__getattr__',
 '__loader__',
 '__name__',
 '__numpy_submodules__',
 '__package__',
 '__path__',
 '__spec__',
 '__version__',
 '_core',
 '_distributor_init',
 '_expired_attrs_2_0',
 '_get_promotion_state',
 '_globals',
 '_int_extended_msg',
 '_mat',
 '_msg',
 '_no_nep50_warning',
 '_pyinstaller_hooks_dir',
 '_pytesttester',
 '_set_promotion_state',
 '_specific_msg',
 '_type_info',
 '_typing',
 '_utils',
 'abs',
 'absolute',
 'acos',
 'acosh',
 'add',
 'all',
 'allclose',
 'amax',
 'amin',
 'angle',
 'any',
 'append',
 'apply_along_axis',
 'apply_over_axes',
 'arange',
 'arccos',
 'arccosh',
 'arcsin',
 'arcsinh',
 'arctan',
 'arctan2',
 'arctanh',
 'argmax',
 'argmin',
 'argpartition',
 'argsort',
 'argwhere',
 'a

In [ ]:
len(dir(np))

532

**out of 532 Built_in_functions we need to learn only Built_in_functions which are used for AI**

- `array `
- `arange `
- `ndim `
- `shape`
- `size `
- `reshape `
- `reshape (-1,1) and reshape (1,-1)`
- `zeros `
- `ones `
- `full `
- `eye `
- `ravel `
- `concatinate `
- `transpose `
- `random -> randint`
- `random -> randn`
- `random -> rand `
- `random -> random `
- `repeat `
- `axis=1`
- `axis = 0 `
- `sum `
- `min `
- `max`
- `argmax `
- `argmin `
- `mat operations `
- `matmul `

In [ ]:
# array -> is used to develop 1D or 2D or 3D etc

a = np.array([10,20,50,90,45])

print(a) # printing 1D values

print(a.shape) # (rows and columns)
print(a.size) # total values

print(a.ndim) # tells how many dimensions

[10 20 50 90 45]
(5,)
5
1


In [ ]:
print(a[2])

50


In [ ]:
# 2D

a = np.array([  [11,4,17]  , [2,8,3] , [5,7,1] ])

print(a) # printing 1D values

print(a.shape) # (rows and columns)
print(a.size) # total values

print(a.ndim) # tells how many dimensions

[[11  4 17]
 [ 2  8  3]
 [ 5  7  1]]
(3, 3)
9
2


In [ ]:
a[1][2]

np.int64(3)

In [ ]:
a[2][0]

np.int64(5)

In [ ]:
# arange = is used to create only 1D data with (n-1) logic

b = np.arange(11,16)
print(b)
print(b.ndim)

[11 12 13 14 15]
1


In [ ]:
b = np.arange(11,19)
print(b)
print(b.ndim)

print('---------------')

c = b.reshape(2,4)
print(c)
print(c.shape)

[11 12 13 14 15 16 17 18]
1
---------------
[[11 12 13 14]
 [15 16 17 18]]
(2, 4)


In [ ]:
# 1D to 2D

b = np.arange(1,61).reshape(10,6)
print(b)
print(b.shape)
print(b.size)

[[ 1  2  3  4  5  6]
 [ 7  8  9 10 11 12]
 [13 14 15 16 17 18]
 [19 20 21 22 23 24]
 [25 26 27 28 29 30]
 [31 32 33 34 35 36]
 [37 38 39 40 41 42]
 [43 44 45 46 47 48]
 [49 50 51 52 53 54]
 [55 56 57 58 59 60]]
(10, 6)
60


In [ ]:
b[4][3]

np.int64(28)

In [ ]:
# 1D to 3D

a = np.arange(1,61).reshape(2,5,6)
a

array([[[ 1,  2,  3,  4,  5,  6],
        [ 7,  8,  9, 10, 11, 12],
        [13, 14, 15, 16, 17, 18],
        [19, 20, 21, 22, 23, 24],
        [25, 26, 27, 28, 29, 30]],

       [[31, 32, 33, 34, 35, 36],
        [37, 38, 39, 40, 41, 42],
        [43, 44, 45, 46, 47, 48],
        [49, 50, 51, 52, 53, 54],
        [55, 56, 57, 58, 59, 60]]])

In [ ]:
a[1][2][4]

np.int64(47)

In [ ]:
a[2][1][2]

np.int64(33)

`reshape function you can convert 1D to 2D -> 2D to 3D anything is possible `

`But reshape(-1,1) and reshape(1,-1) is used to convert 1D to 2D`

In [ ]:
# reshape(-1,1) and reshape(1,-1)

a = np.array([10,20,30,50,90,70])
print(a)
print(a.ndim)
print(a.shape)

print('----------------')
b = a.reshape(-1,1)
print(b)
print(b.shape)
print(b.ndim)

[10 20 30 50 90 70]
1
(6,)
----------------
[[10]
 [20]
 [30]
 [50]
 [90]
 [70]]
(6, 1)
2


In [ ]:
# reshape(-1,1) and reshape(1,-1)

a = np.array([10,20,30,50,90,70])
print(a)
print(a.ndim)
print(a.shape)

print('----------------')
b = a.reshape(1,-1)
print(b)
print(b.shape)
print(b.ndim)

[10 20 30 50 90 70]
1
(6,)
----------------
[[10 20 30 50 90 70]]
(1, 6)
2


In [ ]:
# zeros

a = np.zeros((5,4))
a

array([[0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.]])

In [ ]:
a.dtype

dtype('float64')

In [ ]:
# zeros

a = np.zeros((5,6),dtype = np.int32)
a

array([[0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0]], dtype=int32)

In [ ]:
# ones

a = np.ones((5,6),dtype=np.int32)
a

array([[1, 1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1, 1],
       [1, 1, 1, 1, 1, 1]], dtype=int32)

In [ ]:
# full

a = np.full((5,7) , 11 , dtype = np.float32)
a

array([[11., 11., 11., 11., 11., 11., 11.],
       [11., 11., 11., 11., 11., 11., 11.],
       [11., 11., 11., 11., 11., 11., 11.],
       [11., 11., 11., 11., 11., 11., 11.],
       [11., 11., 11., 11., 11., 11., 11.]], dtype=float32)

In [ ]:
# eye

a = np.eye(3)
a

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [4]:
# ravel = is used to convert any dimension into 1D
a = np.array([  [1,10,5] , [7,6,11] , [5,3,2] ])
print(a)
print(a.ndim)

b = np.ravel(a)
print(b)
print(b.shape)

[[ 1 10  5]
 [ 7  6 11]
 [ 5  3  2]]
2
[ 1 10  5  7  6 11  5  3  2]
(9,)


In [6]:
# Transpose = convert rows to columns and columns to rows

a = np.array([  [1,10,5] , [7,6,11] , [5,3,2] ])
print(a)

print('-------------')
b = np.transpose(a)
print(b)

[[ 1 10  5]
 [ 7  6 11]
 [ 5  3  2]]
-------------
[[ 1  7  5]
 [10  6  3]
 [ 5 11  2]]


In [7]:
# concatination = merging the data

a = np.array([ [10,11] , [12,13] ])
b = np.array([ [100,1] , [200,2] ])

c = np.concatenate((a, b))
print(c)

[[ 10  11]
 [ 12  13]
 [100   1]
 [200   2]]


In [13]:
# numpy -> random -> randint = randomly one integer value will be generated

a = np.random.randint(10,50,3)

print(a)

[18 49 29]


In [16]:
# numpy -> random -> randn = between -3 to 3 random values will come

a= np.random.randn(5 , 4)
a

array([[ 0.48038972, -2.56872406, -1.92149333, -0.3270629 ],
       [ 0.58442291,  1.03610813, -0.59499298, -0.8771491 ],
       [ 0.39470363, -0.87198024,  0.95937963,  1.05941825],
       [ 0.19177503, -1.24119133,  0.73858552, -1.04389296],
       [-0.23407937,  0.4065684 ,  0.71615305,  0.82674843]])

In [18]:
# numpy -> random -> rand = between 0 to 1 random values will come

a= np.random.rand(5 , 4)
a

array([[0.75639263, 0.44212243, 0.79249352, 0.69505941],
       [0.76299516, 0.40095334, 0.26357124, 0.07077657],
       [0.64151434, 0.46055084, 0.46089896, 0.30591552],
       [0.11189641, 0.04843522, 0.45848778, 0.71489877],
       [0.65219507, 0.01540772, 0.97997173, 0.03885636]])

In [19]:
# numpy -> random -> random = between 0 to 1 random values will come

a= np.random.random((5 , 4))
a

array([[0.36023798, 0.03834199, 0.36712191, 0.13973007],
       [0.6324131 , 0.09078526, 0.47975284, 0.37668624],
       [0.8161553 , 0.41168339, 0.85330258, 0.24872418],
       [0.75787277, 0.8568595 , 0.60341066, 0.04716615],
       [0.81206751, 0.46870189, 0.95653913, 0.7842903 ]])

In [20]:
# axis = 1 -> row
# axis = 0 -> column

In [21]:
a = np.array([ [1,10,9] , [7,6,21] , [98,3,2] ])
a

array([[ 1, 10,  9],
       [ 7,  6, 21],
       [98,  3,  2]])

In [22]:
np.sum(a)

np.int64(157)

In [23]:
np.max(a)

np.int64(98)

In [24]:
np.min(a)

np.int64(1)

In [26]:
np.sum(a , axis = 1)

array([ 20,  34, 103])

In [27]:
np.sum(a , axis = 0)

array([106,  19,  32])

In [29]:
np.max(a , axis=0)

array([98, 10, 21])

In [30]:
np.max(a)

np.int64(98)

In [31]:
np.argmax(a)

np.int64(6)

In [32]:
np.min(a)

np.int64(1)

In [33]:
np.argmin(a)

np.int64(0)

In [34]:
np.max(a , axis = 1)

array([10, 21, 98])

In [35]:
np.argmax(a , axis = 1)

array([1, 2, 0])

In [36]:
np.min(a)

np.int64(1)

In [37]:
np.argmin(a)

np.int64(0)

In [38]:
np.min(a , axis = 0)

array([1, 3, 2])

In [39]:
np.argmin(a , axis = 0)

array([0, 2, 2])

In [40]:
a = np.array([ [10,11] , [12,13] ])
b = np.array([ [100,1] , [200,2] ])

np.add(a,b)

array([[110,  12],
       [212,  15]])

In [42]:
np.subtract(a,b)

array([[ -90,   10],
       [-188,   11]])

In [43]:
# matmul

a = np.array( [ [1,2] , [4,3] ])
b = np.array([ [1,4,9] , [3,6,2] ])

np.matmul(a,b)

array([[ 7, 16, 13],
       [13, 34, 42]])

In [44]:
# matmul

a = np.array( [ [1,2] , [4,3] ])
b = np.array([ [1,4,9] , [3,6,2] , [4,6,2] ])

np.matmul(a,b)

ValueError: matmul: Input operand 1 has a mismatch in its core dimension 0, with gufunc signature (n?,k),(k,m?)->(n?,m?) (size 3 is different from 2)

In [47]:
# repeat

a = np.arange(11,16)
b = a.reshape(1, -1)
c = np.repeat(b,10,axis=0)
print(c)

[[11 12 13 14 15]
 [11 12 13 14 15]
 [11 12 13 14 15]
 [11 12 13 14 15]
 [11 12 13 14 15]
 [11 12 13 14 15]
 [11 12 13 14 15]
 [11 12 13 14 15]
 [11 12 13 14 15]
 [11 12 13 14 15]]


In [49]:
b = np.array([ [1,4,9] , [3,6,2] , [4,6,2] ])
b

array([[1, 4, 9],
       [3, 6, 2],
       [4, 6, 2]])

In [51]:
for i in b:
  for j in i:
    if j % 2 != 0:
      print(j)

1
9
3


In [53]:
r = 0

for i in b:
  c = 0
  for j in i:
    if j % 2 != 0:
      print(j)
    c = c + 1

  r = r + 1

print(f'Number of rows : {r}')
print(f'Number of columns : {c}')

1
9
3
Number of rows : 3
Number of columns : 3
